uber/models.py

```diff
@@ -728,21 +729,27 @@ def update_badge(self, attendee, old_badge_type, old_badge_num):
             from uber.badge_funcs import needs_badge_num
 
             if c.SHIFT_CUSTOM_BADGES and c.BEFORE_PRINTED_BADGE_DEADLINE:
-                # fill in the gap from the old number, if applicable
-                badge_num_keep = attendee.badge_num
-                if old_badge_num:
-                    self.shift_badges(old_badge_type, old_badge_num + 1, down=True)
-
-                # make room for the new number, if applicable
+                badge_collision = False
                 if attendee.badge_num:
-                    offset = 1 if old_badge_type == attendee.badge_type and attendee.badge_num > (old_badge_num or 0) else 0
-                    no_gap = self.query(Attendee).filter(Attendee.badge_type == attendee.badge_type,
-                                                         Attendee.badge_num == attendee.badge_num,
-                                                         Attendee.id != attendee.id).first()
-
-                    if no_gap:
-                        self.shift_badges(attendee.badge_type, attendee.badge_num + offset, up=True)
-                attendee.badge_num = badge_num_keep
+                    badge_collision = self.query(Attendee.badge_num).filter(
+                        Attendee.badge_num == attendee.badge_num,
+                        Attendee.id != attendee.id).first()
+                desired_badge_num = attendee.badge_num
+                if old_badge_num:
+                    if attendee.badge_num and badge_collision:
+                        if old_badge_type == attendee.badge_type:
+                            if old_badge_num < attendee.badge_num:
+                                self.shift_badges(old_badge_type, old_badge_num + 1, until=attendee.badge_num, down=True)
+                            else:
+                                self.shift_badges(old_badge_type, attendee.badge_num, until=old_badge_num - 1, up=True)
+                        else:
+                            self.shift_badges(old_badge_type, old_badge_num + 1, down=True)
+                            self.shift_badges(attendee.badge_type, attendee.badge_num, up=True)
```

Comment:
This is the only case where we have to run two badge shift queries:
* The `badge_type` has changed
* There's an `old_badge_number`
* There's a new `badge_number`
* And the new `badge_number` collides with an existing badge number.